File: restApi/web/decisionweak.py
from flask import Flask, request, jsonify
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

app = Flask(__name__)

# Define the thresholds for each subject
thresholds = {
    'Financial Accounting and Reporting': 0.75 * 70,
    'Advanced Financial Accounting and Reporting': 0.75 * 72,
    'Management Services': 0.75 * 72,
    'Auditing': 0.75 * 72,
    'Taxation': 0.75 * 72,
    'Regulatory Framework for Business Transaction': 0.75 * 100
}

# Load historical data (or use your own training data)
data = {
    'Financial Accounting and Reporting': [65, 72, 50, 80, 90],
    'Advanced Financial Accounting and Reporting': [70, 68, 50, 85, 76],
    'Management Services': [72, 85, 50, 80, 65],
    'Auditing': [65, 72, 45, 90, 88],
    'Taxation': [74, 60, 45, 85, 82],
    'Regulatory Framework for Business Transaction': [80, 85, 90, 70, 60]
}

df = pd.DataFrame(data)

# Generate labels based on thresholds
labels = []
for subject, scores in df.items():
    threshold = thresholds.get(subject)
    labels.append([1 if score < threshold else 0 for score in scores])

labels_df = pd.DataFrame(labels).transpose()

# Train the Decision Tree model
X = df
y = labels_df
model = DecisionTreeClassifier(random_state=42)
model.fit(X, y)

@app.route('/predict', methods=['POST'])
def predict():
    student_data = request.json  # Get the student data from the request

    # Convert the student data to a DataFrame for prediction
    student_df = pd.DataFrame([student_data])
    
    # Make prediction using the trained model
    student_predictions = model.predict(student_df)

    # Return the predictions as JSON
    return jsonify(student_predictions.tolist())

if __name__ == '__main__':
    app.run(debug=True)

File: restApi/web/decisionweak.js
const base64 = require('base-64');

try {
    // Read the base64 encoded JSON string from the command line
    const input_data = process.argv[2];

    if (!input_data) {
        throw new Error("No input data provided");
    }

    // Decode the base64 string to JSON format
    const decoded_data = base64.decode(input_data);

    // Parse the JSON string into a JavaScript object
    const student_data = JSON.parse(decoded_data);

    // Define the high scores for each category (adjusted for your requirements)
    const high_scores = {
        'Financial Accounting and Reporting': 70,
        'Advanced Financial Accounting and Reporting': 72,
        'Management Services': 72,
        'Auditing': 72,
        'Taxation': 72,
        'Regulatory Framework for Business Transaction': 100
    };

    // Define the threshold for weakness (score below threshold is considered weak)
    const weakness_threshold = 0.75;  // This represents 75% of the high score

    // Generate predictions based on the threshold and high score for each subject
    function detect_weaknesses(student_data) {
        const predictions = [];
        for (const [subject, score] of Object.entries(student_data)) {
            // Get the high score for the current subject
            const max_score = high_scores[subject] || 100;  // Default to 100 if the subject is not found
            // Calculate the threshold as 75% of the max score
            const threshold_score = max_score * weakness_threshold;
            const int_score = parseInt(score);  // Convert the score to an integer
            // If the student's score is below the threshold, mark it as weak
            if (int_score < threshold_score) {
                predictions.push(1);  // Weak
            } else {
                predictions.push(0);  // Not weak
            }
        }
        return predictions;
    }

    // Get predictions for the student's scores
    const predictions = detect_weaknesses(student_data);

    // Output the predictions as JSON
    console.log(JSON.stringify(predictions));
} catch (error) {
    console.error("Error in decisionweak.js:", error.message);
    // Output a JSON error response
    console.log(JSON.stringify({ error: "An error occurred in decisionweak.js", message: error.message }));
}
